arduino/cores/packagemanager/package_manager.go

```diff
@@ -93,7 +93,11 @@ func (pmb *Builder) BuildIntoExistingPackageManager(target *PackageManager) {
 	target.tempDir = pmb.tempDir
 	target.packagesCustomGlobalProperties = pmb.packagesCustomGlobalProperties
 	target.profile = pmb.profile
-	target.discoveryManager = pmb.discoveryManager
+	if target.discoveryManager != nil {
+		target.discoveryManager.Clear()
+	} else {
+		target.discoveryManager = pmb.discoveryManager
+	}
```

Comment:
I think here:
* `target.discoveryManager` must be always `Clear()` (it cannot be `nil` by construction)
* the discoveries in `pmb.discoveryManager` must be all added to `target.discoveryManager`

```go
	target.discoveryManager.Clear()
	target.discoveryManager.AddAllDiscoveriesFrom(pmb.discoveryManager)
```

I think the only thing you need to implement is the method `AddAllDiscoveriesFrom(...)`.
This would solve also the problem with `areDiscoveriesRunning` because you don't need to track if the discoveries are started: in either case, the original discoverymanager will handle them correctly once they are `Add()`-ed.